File: cryptography/dotnet/Helverify.Cryptography/ZeroKnowledge/ProofOfDecryption.cs
using Helverify.Cryptography.Helper;
using Org.BouncyCastle.Crypto.Digests;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Prng;
using Org.BouncyCastle.Math;
using Org.BouncyCastle.Security;

namespace Helverify.Cryptography.ZeroKnowledge
{
    /// <summary>
    /// Chaum1993
    /// Bernhard2016
    /// KillerProvotum2020
    /// 
    /// Chaum-Pedersen Proof
    /// </summary>
    public class ProofOfDecryption
    {
        private static readonly SecureRandom SecureRandom = new(new DigestRandomGenerator(new Sha256Digest()));

        public BigInteger D { get; }
        public BigInteger U { get; }
        public BigInteger V { get; }
        public BigInteger S { get; }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="d">Proof parameter d</param>
        /// <param name="u">Proof parameter u</param>
        /// <param name="v">Proof parameter v</param>
        /// <param name="s">Proof parameter s</param>
        public ProofOfDecryption(BigInteger d, BigInteger u, BigInteger v, BigInteger s)
        {
            D = d;
            U = u;
            V = v;
            S = s;
        }

        /// <summary>
        /// Creates a proof that the specified ciphertext has been decrypted correctly.
        /// </summary>
        /// <param name="a">First component of an ElGamal ciphertext</param>
        /// <param name="b">Second component of an ElGamal ciphertext</param>
        /// <param name="publicKeyShare">Public key share</param>
        /// <param name="privateKeyShare">Private key share</param>
        /// <returns>Proof of correct decryption</returns>
        public static ProofOfDecryption Create(BigInteger a, BigInteger b, DHPublicKeyParameters publicKeyShare, DHPrivateKeyParameters privateKeyShare)
        {
            BigInteger p = publicKeyShare.Parameters.P;
            BigInteger q = p.Subtract(BigInteger.One).Multiply(BigInteger.Two.ModInverse(p)).Mod(p);
            BigInteger g = publicKeyShare.Parameters.G;
            BigInteger pk = publicKeyShare.Y;
            BigInteger sk = privateKeyShare.X;

            BigInteger r = new BigInteger(privateKeyShare.X.BitLength, SecureRandom).Mod(q);

            BigInteger u = a.ModPow(r, p).Mod(p);
            BigInteger v = g.ModPow(r, p).Mod(p);

            BigInteger c = HashHelper.GetHash(q, pk, a, b, u, v).Mod(q);

            BigInteger s = r.Add(c.Multiply(sk).Mod(q)).Mod(q);

            BigInteger d = a.ModPow(sk, p).Mod(p);

            return new ProofOfDecryption(d, u, v, s);
        }

        /// <summary>
        /// Allows to verify that the specified ciphertext has been decrypted correctly.
        /// </summary>
        /// <param name="a">First component of an ElGamal ciphertext</param>
        /// <param name="b">Second component of an ElGamal ciphertext</param>
        /// <param name="publicKeyShare">Public key share</param>
        /// <returns>True if decryption has been performed correctly, false otherwise.</returns>
        public bool Verify(BigInteger a, BigInteger b, DHPublicKeyParameters publicKeyShare)
        {
            BigInteger p = publicKeyShare.Parameters.P;
            BigInteger q = p.Subtract(BigInteger.One).Multiply(BigInteger.Two.ModInverse(p)).Mod(p);
            BigInteger g = publicKeyShare.Parameters.G;
            BigInteger pk = publicKeyShare.Y.Mod(p);

            BigInteger c = HashHelper.GetHash(q, pk, a, b, U, V).Mod(q);

            bool firstCondition = a.Mod(p).ModPow(S, p).Mod(p).Equals(U.Multiply(D.ModPow(c, p).Mod(p)).Mod(p));

            bool secondCondition = g.Mod(p).ModPow(S, p).Mod(p).Equals(V.Multiply(pk.ModPow(c, p).Mod(p)).Mod(p));

            return firstCondition && secondCondition;
        }
    }
}

File: voter/frontend/src/cryptography/proofOfDecryption.ts
import bigInt from "big-integer";
import {BigNumberHelper} from "../helper/bigNumberHelper";
import {HashHelper} from "../helper/hashHelper";

/**
 * Proves that the decryption of a ciphertext has been performed correctly
 */
export class ProofOfDecryption {

    /**
     * D component
     */
    d: bigInt.BigInteger;

    /**
     * U component
     */
    u: bigInt.BigInteger;

    /**
     * V component
     */
    v: bigInt.BigInteger;

    /**
     * S component
     */
    s: bigInt.BigInteger;

    /**
     * Constructor
     * @param dHex D component as hexadecimal string
     * @param uHex U component as hexadecimal string
     * @param vHex V component as hexadecimal string
     * @param sHex S component as hexadecimal string
     */
    constructor(dHex: string, uHex: string, vHex: string, sHex: string) {
        this.d = BigNumberHelper.fromHexString(dHex);
        this.u = BigNumberHelper.fromHexString(uHex);
        this.v = BigNumberHelper.fromHexString(vHex);
        this.s = BigNumberHelper.fromHexString(sHex);
    }

    /**
     * Verifies that this proof is valid, i.e., that the specified ciphertext has indeed been decrypted correctly
     * @param a C component of the ElGamal ciphertext
     * @param b D component of the ElGamal ciphertext
     * @param pk Public key share of the decrypting consensus node
     * @param p ElGamal parameter p
     * @param g ElGamal parameter g
     */
    verify(a: bigInt.BigInteger, b: bigInt.BigInteger, pk: bigInt.BigInteger, p: bigInt.BigInteger, g: bigInt.BigInteger): boolean {
        const q: bigInt.BigInteger = p.subtract(bigInt.one).multiply(bigInt(2).modInv(p)).mod(p);

        const c: bigInt.BigInteger = HashHelper.getHash(q, [pk, a, b, this.u, this.v]);

        const condition1: boolean = a.mod(p).modPow(this.s, p).mod(p).equals(this.u.multiply(this.d.modPow(c, p).mod(p)).mod(p));

        const condition2: boolean = g.mod(p).modPow(this.s, p).mod(p).equals(this.v.multiply(pk.modPow(c, p).mod(p)).mod(p));

        return condition1 && condition2;
    }
}